elastic/kibana
# [Response Ops][Alerting] Expose shard failures from querying over CCS indices in ES query rule type

Towards https://github.com/elastic/kibana/issues/188466

## Summary

When a query runs over CCS indices and there are shard failures, we see the query return a success response but with an error hidden within the `_shards.failures` or the `_clusters` object. It is possible that some shards/clusters were successful and returned partial results while others failed. There is no indication to the end user that anything is wrong. this PR updates the ES query DSL and KQL rule types to check for any errors hidden within the reponse and report a warning back to the alerting framework. This will put the warning on the rule saved object in the execution status like:

```
{
...
    "execution_status": {
        "status": "warning",
        "warning": {
            "reason": "ruleExecution",
            "message": ": Top hits result window is too large, the top hits aggregator [topHitsAgg]'s from + size must be less than or equal to: [100] but was [300]. This limit can be set by changing the [index.max_inner_result_window] index level setting."
        },
        "last_execution_date": "2024-08-01T20:17:14.771Z",
        "last_duration": 239
    },
....
}
```

which shows up in the UI as:

<img width="1020" alt="Screenshot 2024-08-01 at 4 21 36 PM" src="https://github.com/user-attachments/assets/b3f8a860-b2ab-49e6-a654-9f38d5ff9435">

## To verify

I deployed this PR to a cloud deployment and used it to create a rule that used CCS to search a remote cluster. For the purposes of forcing this error I undid some of the changes from [this PR](https://github.com/elastic/kibana/pull/189291) with [this commit](https://github.com/elastic/kibana/pull/189312/commits/173d093765088a06944dfd29c2e0731130250929) to allow me to create a rule that I know will fail. Then I was able to see the rule running and report back a warning until I fixed the rule config.
